### PR TITLE
Fix problems with/improve warning and invariant rules (take two)

### PR DESCRIPTION
**Pull request by PiPeep**
See the post-merge discussion following #4021. Changes are broken up into three distinct parts to make review easier. Probably the most controversial commit is the last one.

### CODE REVIEW

File: eslint-rules/warning-and-invariant-args.js [hunk 1]
```diff
@@ -58,13 +58,24 @@ module.exports = function(context) {
         );
         return;
       }
-      // count the number of string substitutions, plus the first two args
-      var expectedNArgs = (str.match(/%s/g) || []).length + 2;
+      // If the format formating isn't unique enough. This is also done as a
+      // runtime check in warning's implementations (though not invariant).
```
Comment (spicyj) [inline eslint-rules/warning-and-invariant-args.js]: No longer true?

Comment (PiPeep) [inline eslint-rules/warning-and-invariant-args.js]: Oh, right.

File: src/isomorphic/classic/element/ReactElementValidator.js [hunk 2]
```diff
@@ -97,10 +97,18 @@ function validateExplicitKey(element, parentType) {
   }
   element._store.validated = true;
 
-  warnAndMonitorForKeyUse(
-    'Each child in an array or iterator should have a unique "key" prop.',
-    element,
-    parentType
+  var addendums = getAddendumsForKeyUse('uniqueKey', element, parentType);
```
Comment (spicyj) [inline src/isomorphic/classic/element/ReactElementValidator.js]: https://books.google.com/ngrams/graph?content=addenda%2C+addendums&year_start=1800&year_end=2000&corpus=15&smoothing=3&share=&direct_url=t1%3B%2Caddenda%3B%2Cc0%3B.t1%3B%2Caddendums%3B%2Cc0

File: src/isomorphic/classic/element/ReactElementValidator.js [hunk 3]
```diff
@@ -97,10 +97,18 @@ function validateExplicitKey(element, parentType) {
...
+  if (addendums === null) {
+    // we already showed the warning
+    return;
+  }
+  warning(
+    false,
+    'Each child in an array or iterator should have a unique "key" prop.' +
+    '%s%s%s',
+    addendums.parentOrOwner || '',
```
Comment (spicyj) [inline src/isomorphic/classic/element/ReactElementValidator.js]: Wonder if it's better to merge them into one string first. I guess this is better for analyzing later.

File: src/isomorphic/classic/element/ReactElementValidator.js [hunk 4]
```diff
@@ -128,49 +144,46 @@ function validatePropertyKey(name, element, parentType) {
...
-  var parentOrOwnerAddendum =
-    ownerName ? ` Check the render method of ${ownerName}.` :
-    parentName ? ` Check the React.render call using <${parentName}>.` :
-    '';
+  var addendums = {
+    parentOrOwner:
+      ownerName ? ` Check the render method of ${ownerName}.` :
+      parentName ? ` Check the React.render call using <${parentName}>.` :
+      null,
+    url: ' See https://fb.me/react-warning-keys for more information.',
```
Comment (spicyj) [inline src/isomorphic/classic/element/ReactElementValidator.js]: can you add `childOwner: null,` here for consistency?

**Pull request merged**